Amd may16 2025 support freertos

bentheredonethat opened the pull request:
@arnopo  here is new attempt at bringing in freertos support for legacy apps
CC @tnmysh

tnmysh commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,84 @@
+#include "FreeRTOS.h"
+#include "task.h"
+
+#define LPRINTF(format, ...) xil_printf(format, ##__VA_ARGS__)
```
use metal_err. Please take care of this for each PR you have to open from here on. We don't really need any other way to print logs. Only metal_* macros.

bentheredonethat replied:
will fix.

tnmysh commented on examples/legacy_apps/examples/rpc_demo/CMakeLists.txt:
```diff
@@ -26,6 +26,13 @@ endif (${PROJECT_SYSTEM} STREQUAL "linux")
   list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/${_app}.c")
+  if (${_app} STREQUAL rpc_demo)
+    if (${PROJECT_SYSTEM} STREQUAL "linux")
+        list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/generic/main.c")
```
Arnaud had pointed this before. What is importance of generic/main.c under Linux OS? Can you please check which demo is using this option?

bentheredonethat replied:
as noted below, Linux remote will use this. confirmed Linux build worked for this.

bentheredonethat replied:
added comment in latest PR for this.

arnopo replied:
the term  generic is used for the baremetal in libmetal. and this approach could generate some regression for non Linux system.

Regarding this commit I wonder if we should not just keep `main()` in the example using `weak`  attribute and here append `${PROJECT_SYSTEM}/main.c `if the file exists. The  ${PROJECT_SYSTEM}/main.c` would redefine the `main()`

tnmysh commented on examples/legacy_apps/examples/matrix_multiply/CMakeLists.txt:
```diff
@@ -16,6 +16,13 @@ set (OPENAMP_LIB open_amp)
   list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/${_app}.c")
+  if (${_app} STREQUAL matrix_multiplyd)
+    if (${PROJECT_SYSTEM} STREQUAL "linux")
+      list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/generic/main.c")
```
Please check which linux demo is using this option and if it's needed, then please add comment explaining it.

bentheredonethat replied:
the OpenAMP Remote will be using this (rpmsg-echo,  matrix_multiplyd and rpc_demo)
Linux build was tested to work.

arnopo commented on examples/legacy_apps/examples/echo/generic/main.c:
```diff
@@ -0,0 +1,65 @@
+#include <metal/log.h>
+#include <metal/version.h>
+#include "platform_info.h"
+#include "rpmsg-echo.h"
```
Please reorder using alphabetical order.

arnopo commented on examples/legacy_apps/examples/echo/generic/main.c:
```diff
@@ -0,0 +1,65 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
```
copyright is missing

bentheredonethat replied:
will fix.

arnopo commented on examples/legacy_apps/examples/echo/generic/main.c:
```diff
@@ -0,0 +1,65 @@
+#define LPERROR(fmt, ...) metal_err("ERROR: " fmt, ##__VA_ARGS__)
+
+/* Application entry point */
+int main(int argc, char *argv[])
```
I would expect that this function (same for other examples) is just a cut/past of the one in `rpmsg-echo.c`

bentheredonethat replied:
will fix.

arnopo commented on examples/legacy_apps/examples/echo/CMakeLists.txt:
```diff
@@ -21,6 +21,13 @@ foreach (_app rpmsg-echo-ping rpmsg-echo)
+      list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/generic/main.c")
+    else()
+      list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/${PROJECT_SYSTEM}/main.c")
+    endif(${PROJECT_SYSTEM} STREQUAL "linux")
```
@wmamills : does this generate some regression in your CI?

arnopo commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -46,6 +50,11 @@ static int rpmsg_endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
 		LPERROR("rpmsg_send failed\r\n");
 	}
+
+#ifdef FREERTOS_BSP
```
is it a standard config or an AMD config?

arnopo commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -46,6 +50,11 @@ static int rpmsg_endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
 	}
+
+#ifdef FREERTOS_BSP
+	xTaskResumeFromISR(rpmsg_task);
```
I did not see that the task was suspended. Do I missed something?
This function should be generic if needed. with an implementation per system

arnopo commented on examples/legacy_apps/system/freertos/machine/zynqmp_r5/gic_init.c:
```diff
@@ -0,0 +1,32 @@
+/*
```
Please Split this commit, one to introduce free RTOS, one for AMD implementation

bentheredonethat commented:
Hi @arnopo  i have updated the PR as follows:

1. split generic rename demo logic and GIC init
2. split freertos demos and freertos GIC init
3. add stubs that make app-suspend for freertos suspend/resume generic 

style fixes

arnopo commented on examples/legacy_apps/system/freertos/machine/zynqmp_r5/gic_init.c:
```diff
@@ -0,0 +1,32 @@
+#include "xscugic.h"
+
+/* Interrupt Controller setup */
+int app_gic_initialize(void)
```
Would it be possible to move this in example/legacy_apps/ machine? 
Would be nice that the system folder does not contain machine specificities.
This request is mainly for the FreeRTOS support that is new .

arnopo commented on examples/legacy_apps/system/freertos/machine/zynqmp_r5/app_task.c:
```diff
@@ -0,0 +1,24 @@
+
+void app_suspend(void)
+{
+vTaskSuspend(rpmsg_task);
```
tabulation

arnopo commented on examples/legacy_apps/system/freertos/machine/zynqmp_r5/app_task.c:
```diff
@@ -0,0 +1,24 @@
+{
+	xTaskResumeFromISR(rpmsg_task);
+}
+#endif /* FREERTOS_BSP */
```
seems to me that is file could be generic changing function name and passing task as 
 a void * argument

arnopo reviewed (commented):
Much better, thanks for the split.
The last point I would like to improve is to avoid to have machine specificity ( if possible) in the `system/freertos`
I would expect that routine in this folder is just a wrapper to Freertos API

bentheredonethat commented:
Hi @arnopo  i updated so that hte gic_init logic is (a) only there for r5 and (b) i have used format similar to earlier suggestion where the gic_init and app-resume logic specific to each OS is only there in the zynqmp_r5 area.

bentheredonethat commented:
@arnopo  i updated the app-task logic to be generic/same for freertos and baremetal... but im not sure its better than it was before because we still have to do an extern to expose the rpmsg-task wherever the suspend/resume APIs are called.

arnopo commented:
> > rpmsg-task
> 
> What about adding rmpsg_task_suspend/resume(void) function in main in main.c that will call vTaskSuspend/Resume? and creating a main.h that would be include in the machine if neded to suspend/resume the FreeRTOS rpmsg_task?

@bentheredonethat 
I don't see your last comment received by mail, so I add it here too answer

>>What about adding rmpsg_task_suspend/resume(void) function in main in main.c that will call vTaskSuspend/Resume? and creating a main.h that would be include in the machine if neded to suspend/resume the FreeRTOS rpmsg_task?


> Hi @arnopo
>i'm missing something here - could you explain this?

Yes sorry it is not cristal clear. The main principle is that the  suspend/resume of the `rpmsg_task` task seems to be generic.
Regarding your implementation you need to call it in zynqmp_r5_a53_rproc.c while you call `_rproc_wait` in zynqmp_r5/platform_info.c

As you mention the previous implementation I proposed implies dependency between the sample and the platform to export  `rpmsg_task`
 
>
> the catch here is that the platform_poll() and zynqmp_r5_a53_proc_irq_handler() routines need to have the rpmsg_task_suspend/resume calls.
> If I understand your suggestion is it to:
> (a) define main.h with weak implementations

add follwing function in main.c
```
void rpmsg_app_suspend(void)
{
	vTaskSuspend(rpmsg_task);
}

void rpmsg_app_resume(void)
{
	xTaskResumeFromISR(rpmsg_task);
}
```
Create a main.h that declare these functions

> (b) keep commit legacy_apps: zynqmp_r5: Add default usage of app_suspend/resume but it will now also include main.h

- come back to the call of `_rproc_wait` in `zynqmp_r5/platform_info.c`
- include the main.h and call r`pmsg_app_suspend`/`rpmsg_app_resume` in `zynqmp_r5_a53_rproc.c `

> (c) define new freertos-amd specific implementations in the freertos main.c files for each demo
> just want to make sure i understand the suggestion here
> 

I hope this explanation clarifies things. Don't hesitate to reach out if anything is still unclear or if you see an issue with my proposal.

bentheredonethat commented:
Hi @arnopo  apprecate you the time to explain here.

The one concern I have is that we have 3 different main routines per OS. So I would be adding that same freertos implementation to the 3 legacy apps' freertos main.c files and likewise for generic main.c files.

That is why I added the implementaion in separate file, so that the implementation is not duplicated.

If you still want me to proceed in doing the implementation in each of the legacy app' s main.c files I will do so with a 'weak' attribute and test that.

thank you!

arnopo commented:
> Hi @arnopo apprecate the time to explain here.

Thank you for taking the time to consider my suggestions. 😃

> 
> The one concern I have is that we have 3 different main routines per OS. So I would be adding that same freertos implementation to the 3 legacy apps' freertos main.c files and likewise for generic main.c files.

Your point is valid !

- For the generic main is seems not needed as you call `_rproc_wait` in your routines, so you could come back to your initial implementation removing app_suspend/resume, right?
- For the FreeRTos: yes this this the draw back. from my point of view i would prefer to duplicate than hide OS dependent code in the machine.

What about creating a examples/legacy_apps/system/freertos/thread.c file to communalise the routine that could be in future reused for some other OSes? 
- Add  `thread_TaskCreate function (task_ID, task_func_ , task_name, stack_size, priority, p_created_task)`
    => the `task_ID` would be an index of  a table of static TaskHandle_t tasks, and defined as an enum in thread.h
```
    TaskHandle_t  task_kist[THREAD_APP_NB_TASK_ID];
```
- Add `thread_task_suspend` and `thread_task_resume` functions with task ID provided as argument.
- create a thread.h 
   - containing thread routine declarations
   - containing an enum that list task IDs 
```
enum {
    THREAD_APP_TASK_ID = 0,
    THREAD_APP_NB_TASK_ID,
} thread_task_id;
``` 

> 
> That is why I added the implementaion in separate file, so that the implementation is not duplicated.
> 
> If you still want me to proceed in doing the implementation in each of the legacy app' s main.c files I will do so with a 'weak' attribute and test that.
> 
> thank you!

edmooring reviewed (approved):
The amount of duplicated code in this PR bothers me, but I don't see an easy way to refactor it at the moment. Since these are legacy apps, and therefore unlikely to be modified, I think we can let it go.

arnopo commented on examples/legacy_apps/examples/rpc_demo/generic/main.c:
```diff
@@ -25,6 +26,18 @@
+void rpmsg_app_suspend(void *data)
+{
+	(void)data;
+	_rproc_wait();
```
`_rproc_wait` is AMD specific ,  it should not be call here, 

I would make this function empty or just remove `rpmsg_app_suspend`and  `rpmsg_app_resume` for generic implementation if unused

bentheredonethat commented:
Hi @arnopo  you mentioned 
"_rproc_wait is AMD specific , it should not be call here,

I would make this function empty or just remove rpmsg_app_suspendand rpmsg_app_resume for generic implementation if unused"

and 

"If rpmsg_app_suspend and rpmsg_app_resume are removed for generic, this commit become useless"


thanks for calling these out. we use ARM instruction wfi here in _rproc_wait.

it is small but _rproc_wait is used in this case because of that instruction.

Also I might be missing something here -- but the _rproc_wait is in fact replaced in a later commit by rpmsg_app_suspend so it does 'go away'. Unless this coment is calling something else out here?

I agree with the sentimenet that these routines are small and in the generic context it makes sense to keep them as is.

BUT the real utility of those comes in only for the freertos implementation. The generic implementation is there for consistency and convention.

Given that context, please advise on what to change here, I am unsure of the action t for this suspend/resume part on this PR.

arnopo commented:
> Hi @arnopo you mentioned "_rproc_wait is AMD specific , it should not be call here,
> 
> I would make this function empty or just remove rpmsg_app_suspendand rpmsg_app_resume for generic implementation if unused"
> 
> and
> 
> "If rpmsg_app_suspend and rpmsg_app_resume are removed for generic, this commit become useless"
> 
> thanks for calling these out. we use ARM instruction wfi here in _rproc_wait.
> 
> it is small but _rproc_wait is used in this case because of that instruction.
> 
> Also I might be missing something here -- but the _rproc_wait is in fact replaced in a later commit by rpmsg_app_suspend so it does 'go away'. Unless this coment is calling something else out here?
> 
> I agree with the sentimenet that these routines are small and in the generic context it makes sense to keep them as is.
> 
> BUT the real utility of those comes in only for the freertos implementation. The generic implementation is there for consistency and convention.
> 
> Given that context, please advise on what to change here, I am unsure of the action t for this suspend/resume part on this PR.

My bad [examples/legacy_apps/machine/zynqmp_r5/platform_info.c](https://github.com/OpenAMP/openamp-system-reference/pull/76/files#diff-1f7c261e2b925054f236b7efc0d6ed75209fbb4054c955145c13135b327643f0) is comon for freeRtos and generic

I implemented a solution that should work. At least it works on my Linux machine ( demo on PC)
Could you test it and tell me if it is ok for you?
My branch based on you PR: https://github.com/arnopo/openamp-system-reference/tree/app_suspend
if ok for you fill free to update your PR with that ( fixing potential bugs)

bentheredonethat commented:
Hi @arnopo  thanks for the collaboration here! LGTM on those 2 patches. see updated branch below.

arnopo commented on examples/legacy_apps/machine/zynqmp_r5/platform_info.c:
```diff
@@ -54,8 +55,7 @@
 
-#ifndef RPMSG_NO_IPI
-#define _rproc_wait() asm volatile("wfi")
+
```
Remove added blanck line

arnopo commented on examples/legacy_apps/machine/zynqmp/platform_info.c:
```diff
@@ -108,9 +108,6 @@ static struct remoteproc rproc_inst;
 extern int init_system(void);
```
If you squash the patches in one please also squash the commit removing 
```
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

examples: legacy: Add system suspend/resume
```
in the middle of the commit message

arnopo reviewed (approved):
2 minor comments, else LGTM

bentheredonethat commented:
@arnopo  udpated!

arnopo commented:
@edmooring : could you have a look please

edmooring reviewed (approved):
Looks good to go.
